### PR TITLE
Issue/lineno

### PR DESCRIPTION
**Pull request by bartv**
Linenumber fixes
1. Add asserts to ensure that linenumber 0 is no longer allowed
2. Locations that are actually used cannot be None
3. Fix attach lnr indexes for a number of parser rules

### CODE REVIEW

File: src/inmanta/ast/__init__.py [hunk 1]
```diff
@@ -67,10 +67,18 @@ def __str__(self) -> str:
 class Locatable(object):
 
     def __init__(self):
-        self.location = None  # type: Location
+        self._location = None  # type: Location
+
+    def set_location(self, location: Location):
+        assert location is not None
+        assert location.file == "internal" or location.lnr > 0
```
Comment (wouterdb) [inline src/inmanta/ast/__init__.py]: perhaps modify the "internal" lines to also be >0, that number is arbitrary anyway 
this code is executed quite often so the additional string compare can be avoided

File: src/inmanta/ast/statements/generator.py [hunk 2]
```diff
@@ -48,6 +48,7 @@ class SubConstructor(GeneratorStatement):
     def __init__(self, instance_type: "", implements: "Implement") -> None:
         GeneratorStatement.__init__(self)
         self.type = instance_type
+        self.location = instance_type.location
```
Comment (wouterdb) [inline src/inmanta/ast/statements/generator.py]: perhaps the location of the implements would be better?

Comment (wouterdb) [reply src/inmanta/ast/statements/generator.py]: (consistent with line 63)

File: src/inmanta/ast/__init__.py (hunk 1 above)
Comment (bartv) [reply src/inmanta/ast/__init__.py]: Could be useful, not sure yet if we keep these asserts or not

**Comment by wouterdb** [comment]
two questions remain 
1- should we switch to PLY line tracking (quite some work, result is uncertain, but might reduce the amount of code)
2- should we switch all source locations from linenr to character range (a lot of work, what the character range means is not always clear, may result in more uniform code)

At this point, I would not address the two issues, but we should keep them in mind.

**Comment by bartv** [comment]
Indeed. I leave that up to you. This main reason for this branch is to once and for all make sure that everything has a linenumber and thus a meaningful error

**Pull request merged**